### PR TITLE
Add support for Dell PowerProtect Data Manager

### LINKED ISSUE #3126: Implement BACKUP=PPDM for Dell PowerProtect Data Manager
Sponsored implementation for https://www.dell.com/en-us/dt/data-protection/powerprotect-data-manager.htm

### PR DESCRIPTION
**Pull request by schlomo**
#### Relax-and-Recover (ReaR) Pull Request Template

Please fill in the following items before submitting a new pull request:

##### Pull Request Details:

* Type: **New Feature**

* Impact: **Normal** 

* Reference to related issue (URL): Fixes #3126 

* How was this pull request tested?

Manual tests in Dell lab environment on RHEL/compatible 7,8,9 and SLES 12,15 and Ubuntu 20,22

* Description of the changes in this pull request:

Add support for PPDM and fix bugs we encounted.

Introduces Bash associative arrays, mandating Bash 4

### CODE REVIEW

**Comment by schlomo** [comment]
@rear/contributors Please have a look, I'd like to do the last fixes on Monday and merge because that is the limit for the Dell lab where I can run tests.

**Comment by pcahyna** [comment]
First note: please fix the commit message of f15344afe91ac07b6015563f917dc7b141519bed and force-push.

**Comment by pcahyna** [comment]
Second impression: please do not mix formatting changes and actual functionality changes in the same commit. 24c122798406e0d62887e0e2909bb942d44a88dd is about "Ensure minimum Bash version ", but it contains hundreds of unrelated changed lines.
I also have the impression that many of the formatting changes should be avoided, as they make toie code look worse (and less in line with the examples in https://github.com/rear/rear/wiki/Coding-Style ).

**Comment by schlomo** [comment]
Thanks @pcahyna for helping us to be better contributors, you are right in both points.

**Comment by schlomo** [comment]
Im actually unhappy with 
https://github.com/rear/rear/pull/3138/files#diff-0cf7e0a3b217a0feaf68c285ada67dbf21b981b3c6525c79d3fc29b9d956ada8R98-R103

because it mixes in all the backup methods into a single file instead of using our directory based overlay architecture to separate the concerns for each backup software. If it is OK with @rear/contributors then I'd gladly pull this code apart by introducing a new `LD_LIBRARY_PATH_ADDITIONS` variable that can be set for every backup tool.

The reason why I didn't do that is that I cannot test this change for all backup tools so that we would have to rely on eyeball inspection and later tests by others.

**Comment by schlomo** [comment]
Another thought: I added a feature to backup and restore the ESP because PPDM doesn't backup `vfat` filesystems. I'm wondering if I should change this code to be more generic so that it would either happen automatically for any recovery that doesn't restore the ESP or that it can be activated for those commercial backup tools that can't backup the ESP themselves.

In my test case the ESP was ~7MB, not much at all.

**Comment by pcahyna** [comment]
Would it work to add a PPDM-specific check that there are not vfat filesystems in the layout (except the ESP)?

File: usr/share/rear/lib/shell-workflow.sh [hunk 1]
```diff
@@ -1,13 +1,14 @@
 if [[ "$VERBOSE" ]]; then
     WORKFLOW_shell_DESCRIPTION="start a bash within rear; development tool"
 fi
-WORKFLOWS+=( shell )
-WORKFLOW_shell () {
-    if test "$SIMULATE" ; then
+WORKFLOWS+=(shell)
+WORKFLOW_shell() {
+    if test "$SIMULATE"; then
```
Comment (pcahyna) [inline usr/share/rear/lib/shell-workflow.sh:6]: please edit the commit to remove this unrelated whitespace change (which is contrary to examples in https://github.com/rear/rear/wiki/Coding-Style as well )

File: usr/share/rear/lib/rear-shell.bashrc [hunk 2]
```diff
@@ -20,7 +22,7 @@ unset REAR_EVAL # reduce environmental pollution
 VERBOSE=1
 
 # source Relax-and-Recover functions
-for script in $SHARE_DIR/lib/*functions.sh ; do source $script ; done
+for script in $SHARE_DIR/lib/*functions.sh; do source $script; done
```
Comment (pcahyna) [inline usr/share/rear/lib/rear-shell.bashrc:25]: please edit the commit to remove this unrelated whitespace change (which is contrary to examples in https://github.com/rear/rear/wiki/Coding-Style as well )

**Comment by pcahyna** [comment]
> Thanks @pcahyna for helping us to be better contributors, you are right in both points.

thanks for the update, it looks much better now, but there are still some places with unrelated white space changes remaining. Please also drop the huge autogenerated html files from the PR, unless they have been in the source tree before.

File: usr/share/rear/verify/PPDM/default/390_check_for_running_ppdm_agent_service.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
+let count=0
```
Comment (pcahyna) [inline usr/share/rear/verify/PPDM/default/390_check_for_running_ppdm_agent_service.sh:1]: do you need `let` and later `unset` ? Wouldn't `local` be better in this case?

File: usr/share/rear/verify/PPDM/default/400_check_and_wait_for_valid_lockbox.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        return 1
+    fi
+    Log "lockbox result:$LF$res"
+
+    read -d '' _h1 _h2 _h3 _h4 PPDM_DD_IP PPDM_STORAGE_UNIT PPDM_DD_USERNAME PPDM_DD_CONFIG_TYPE <<<"$res"
+    [[ "$_h1" == DD_IP && "$_h2" == Storage_Unit && $_h3 == DD_USERNAME && $_h4 == DD_Config_Type ]] ||
+        BugError "Headings of ddfsadmin lockbox -query differ from what we expect: $_h1 $_h2 $_h3 $_h4"
+}
+
+let start_wait_for_lockbox=$SECONDS waiting=0 found=0
```
Comment (pcahyna) [inline usr/share/rear/verify/PPDM/default/400_check_and_wait_for_valid_lockbox.sh:28]: can you use `local` instead of `let` and `unset` ?

File: usr/share/rear/verify/PPDM/default/410_find_backup_sets.sh [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+# The danger of this approach is that ReaR would also restore file systems that have been removed from a server,
+# and this can of course make the recovery fail if there is not enough disk space available for it.
+#
+# We believe that the latter case will happen much more seldom and that the benefit of restoring a file system
+# that wasn't saved in the latest backup run by far outweighs the danger.
+
+local res= backup_info= pit_timestamp= header=1 # start from parsing the header
+
+res=$(ddfsadmin backup query -remote -d "$PPDM_DD_IP" -s "$PPDM_STORAGE_UNIT" -u "$PPDM_DD_USERNAME" 2>&1)
+test "$?" -eq 0 || Error "Could not query PPDM backups:$LF$res"
```
Comment (pcahyna) [inline usr/share/rear/verify/PPDM/default/410_find_backup_sets.sh:50]: You can combine the previous two lines:
```
res=$(ddfsadmin backup query -remote -d "$PPDM_DD_IP" -s "$PPDM_STORAGE_UNIT" -u "$PPDM_DD_USERNAME" 2>&1) || Error "Could not query PPDM backups:$LF$res"
```

File: usr/share/rear/verify/PPDM/default/400_check_and_wait_for_valid_lockbox.sh [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+# DD_IP           Storage Unit                                                     DD_USERNAME          DD_Config_Type
+# 192.168.1.30    Linux-ppdm-01-f3392/PLCTLP-580da36a-9e5d-4bc3-a438-d7d43cf34da4  Linux-ppdm-01-f3392  PROTECTION
+
+# failure return code also 0 and this or other output
+# 2024-01-16T15:30:51.103Z Error retrieving password from lockbox
+
+function ppdm_get_config_from_lockbox {
+    # Note: We replace Storage Unit with Storage_Unit so that our read call below can successfully split on space characters :-)
+    local res=$(ddfsadmin lockbox -query 2>&1 | sed -e 's/Storage Unit/Storage_Unit/g') _h1 _h2 _h3 _h4
+    (($? > 0)) && Error "Could not query lockbox:$LF$res"
```
Comment (pcahyna) [inline usr/share/rear/verify/PPDM/default/400_check_and_wait_for_valid_lockbox.sh:15]: this won't work because the exit code from the previous line is the exit code from `local`, not from `ddfsadmin`.
Suggestion:
```
    local res
    res=$(ddfsadmin lockbox -query 2>&1 | sed -e 's/Storage Unit/Storage_Unit/g') _h1 _h2 _h3 _h4 || Error "Could not query lockbox:$LF$res"
```

File: usr/share/rear/restore/PPDM/default/410_restore_esp.sh [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Restore the ESP from ReaR rescue image
+
+[[ -d /mnt/local/boot/efi && -d /boot/efi ]] || return
+
+local espfiles=$(echo /mnt/local/boot/efi/*)
```
Comment (pcahyna) [inline usr/share/rear/restore/PPDM/default/410_restore_esp.sh:5]: cosmetic: https://www.shellcheck.net/wiki/SC2155

**Comment by schlomo** [comment]
@pcahyna about the white space changes like `do source $script;` vs. `$script ;` I have the challenge that there doesn't seem to be a [shfmt](https://github.com/mvdan/sh/blob/master/cmd/shfmt/shfmt.1.scd) setting to allow that.

I really would like us to find a way to use automated formatting, as it significantly reduces mental overhead for us developers. BTW, The [Google Shell Guide](https://google.github.io/styleguide/shellguide.html) also doesn't require a blank before the `;` in a `do` statement, so maybe we don't need to require that.

What I would specifically suggest is to amend our `.editorconfig` file like this:
```
root = true
[*]
end_of_line = lf
insert_final_newline = true
indent_style = space
indent_size = 4
shell_variant = bash
binary_next_line = true
switch_case_indent = true
space_redirects = true
keep_padding = true
function_next_line = false

[[bash]]
shell_variant = bash
```

and be happy what `shfmt` makes out of it.

**Comment by schlomo** [comment]
about the HTML files: I added them to our `.gitignore` so that future runs of `make doc` won't have this effect

**Comment by schlomo** [comment]
BTW, I'm now reworking the `LD_LIBRARY_PATH` code to be modularized and the `/boot/efi` code to be reusable.

File: .gitignore [hunk 8]
```diff
@@ -11,3 +11,4 @@ var/log/rear
 .vscode
 servers
 /tools/run-in-docker-script.sh
+*.html
```
Comment (pcahyna) [inline .gitignore:14]: can you please split this change into its own commit as it is unrelated to "Add support for new backup software Dell PowerProtect Data Manager"?

**Comment by pcahyna** [comment]
> I also found another false positive for the `systemd` libraries which is apparently caused by the same problem (`ldd` not picking up libraries from the current working directory):
> 
> ```
> Testing each binary with 'ldd' and look for 'not found' libraries within the recovery system
> /usr/lib64/systemd/libsystemd-core-252.so requires additional libraries
>         libsystemd-shared-252.so => not found
> ```

this message is harmless, annoying though. There is an issue about it already : #3021

File: .gitignore (hunk 8 above)
Comment (schlomo) [reply .gitignore:14]: Done

File: doc/rear.8 [hunk 9]
```diff
@@ -31,7 +31,7 @@
 rear \- bare metal disaster recovery and system migration tool
 .SH "SYNOPSIS"
 .sp
-\fBrear\fR [\fB\-h\fR|\fB\-\-help\fR] [\fB\-V\fR|\fB\-\-version\fR] [\fB\-dsSv\fR] [\fB\-D\fR|\fB\-\-debugscripts\fR \fISET\fR] [\fB\-c\fR \fIDIR\fR] [\fB\-C\fR \fICONFIG\fR] [\fB\-r\fR \fIKERNEL\fR] [\-\-] \fICOMMAND\fR [\fIARGS\fR\&...]
+\fBrear\fR [\fB\-h\fR|\fB\-\-help\fR] [\fB\-V\fR|\fB\-\-version\fR] [\fB\-dsSv\fR] [\fB\-D\fR|\fB\-\-debugscripts\fR \fISET\fR] [\fB\-c\fR \fIDIR\fR] [\fB\-C\fR \fICONFIG\fR] [\fB\-r\fR \fIKERNEL\fR] [\fB\-n\fR|\fB\-\-non\-interactive\fR] [\fB\-e\fR|\fB\-\-expose\-secrets\fR] [\-\-] \fICOMMAND\fR [\fIARGS\fR\&...]
```
Comment (schlomo) [inline doc/rear.8:34]: BTW, this shows that after adding the `-e` switch nobody ran `make doc` :shrug:

**Comment by schlomo** [comment]
Yes, let's move the coding style change out of here to #3142

**Comment by schlomo** [comment]
@pcahyna I think I fixed everything, can we now talk about PPDM?

File: doc/rear.8 (hunk 9 above)
Comment (pcahyna) [reply doc/rear.8:34]: Yes @schlomo good observation, so I don't object to committing this file after you have regenerated it, despite having seemingly unrelated changes.
I hate having this autogenerated file in the source tree btw. As there is a Make target to generate it, it should be generated during build, IMO. Its presence in sources makes the build undeterministic: if the Git checkout has older rear.8 than rear.adoc, it will get regenerated during the build, but if it is newer, it will not be generated (masking possible errors). But this discussion belongs elsewhere.

Comment (schlomo) [reply doc/rear.8:34]: I can promise to amend our `.gitignore` and remove all build results after this PR is done.

**Comment by schlomo** [comment]
If there are no more concerns then I'll merge tomorrow morning and continue with the other topics.

**Comment by pcahyna** [comment]
@schlomo thanks for cleaning up the commit history, looks good now.
I have not had a chance to review the latest changes (generalized ESP restore and LD_LIBRARY_PATH stuff). I have some minor comments about them that can be addressed in a separate PR.
